File: app/src/main/java/com/salman/bitclock/data/models/Alarm.java
package com.salman.bitclock.data.models;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

import java.util.Calendar;
import java.util.Objects;

@Entity(tableName = "alarms")
public class Alarm {

    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    private int id;

    @ColumnInfo(name = "hour")
    private int hour;

    @ColumnInfo(name = "minute")
    private int minute;

    @ColumnInfo(name = "is_enabled")
    private boolean enabled;

    @ColumnInfo(name = "label")
    private String label;

    @ColumnInfo(name = "repeat_days")
    private int repeatDays; // Bitmask for repeating days

    @ColumnInfo(name = "snooze_duration")
    private int snoozeMinutes;

    @ColumnInfo(name = "sound_uri")
    private String soundUri;

    @ColumnInfo(name = "vibrate")
    private boolean vibrate;

    @ColumnInfo(name = "created_at")
    private long createdAt;

    // Getters and Setters

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public int getHour() {
        return hour;
    }

    public void setHour(int hour) {
        this.hour = hour;
    }

    public int getMinute() {
        return minute;
    }

    public void setMinute(int minute) {
        this.minute = minute;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public String getLabel() {
        return label;
    }

    public void setLabel(String label) {
        this.label = label;
    }

    public int getRepeatDays() {
        return repeatDays;
    }

    public void setRepeatDays(int repeatDays) {
        this.repeatDays = repeatDays;
    }

    public int getSnoozeMinutes() {
        return snoozeMinutes;
    }

    public void setSnoozeMinutes(int snoozeMinutes) {
        this.snoozeMinutes = snoozeMinutes;
    }

    public String getSoundUri() {
        return soundUri;
    }

    public void setSoundUri(String soundUri) {
        this.soundUri = soundUri;
    }

    public boolean isVibrate() {
        return vibrate;
    }

    public void setVibrate(boolean vibrate) {
        this.vibrate = vibrate;
    }

    public long getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(long createdAt) {
        this.createdAt = createdAt;
    }

    // Helper methods

    public boolean isRepeating() {
        return repeatDays != 0;
    }

    public long getNextAlarmTime() {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, hour);
        calendar.set(Calendar.MINUTE, minute);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        if (calendar.getTimeInMillis() <= System.currentTimeMillis()) {
            calendar.add(Calendar.DAY_OF_YEAR, 1);
        }

        if (!isRepeating()) {
            return calendar.getTimeInMillis();
        }

        for (int i = 0; i < 7; i++) {
            int dayOfWeek = calendar.get(Calendar.DAY_OF_WEEK); // SUNDAY = 1, MONDAY = 2, ...
            int dayBit = 1 << ((dayOfWeek + 5) % 7); // Convert to Mon-Sun bitmask

            if ((repeatDays & dayBit) != 0) {
                return calendar.getTimeInMillis();
            }
            calendar.add(Calendar.DAY_OF_YEAR, 1);
        }
        return 0; // Should not happen
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Alarm alarm = (Alarm) o;
        return id == alarm.id &&
                hour == alarm.hour &&
                minute == alarm.minute &&
                enabled == alarm.enabled &&
                repeatDays == alarm.repeatDays &&
                snoozeMinutes == alarm.snoozeMinutes &&
                vibrate == alarm.vibrate &&
                createdAt == alarm.createdAt &&
                Objects.equals(label, alarm.label) &&
                Objects.equals(soundUri, alarm.soundUri);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, hour, minute, enabled, label, repeatDays, snoozeMinutes, soundUri, vibrate, createdAt);
    }
}

File: app/src/main/kotlin/com/salman/bitclock/data/models/Alarm.kt
package com.salman.bitclock.data.models

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import java.time.LocalDateTime
import java.time.ZoneId

enum class MissionType {
    NONE, MATH, SHAKE, BARCODE
}

@Entity(tableName = "alarms")
data class Alarm(
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    val id: Int = 0,

    @ColumnInfo(name = "hour")
    val hour: Int,

    @ColumnInfo(name = "minute")
    val minute: Int,

    @ColumnInfo(name = "is_enabled")
    val isEnabled: Boolean = true,

    @ColumnInfo(name = "label")
    val label: String = "",

    @ColumnInfo(name = "repeat_days")
    val repeatDays: Int = 0, // Bitmask for repeating days (Mon=1, Tue=2, ..., Sun=64)

    @ColumnInfo(name = "snooze_duration")
    val snoozeMinutes: Int = 10,

    @ColumnInfo(name = "sound_uri")
    val soundUri: String = "",

    @ColumnInfo(name = "vibrate")
    val isVibrate: Boolean = true,

    @ColumnInfo(name = "created_at")
    val createdAt: Long = System.currentTimeMillis(),

    @ColumnInfo(name = "mission_type")
    val missionType: MissionType = MissionType.NONE,

    @ColumnInfo(name = "mission_difficulty")
    val missionDifficulty: Int = 1,

    @ColumnInfo(name = "mission_target")
    val missionTarget: String = "",

    @ColumnInfo(name = "snooze_limit")
    val snoozeLimit: Int = 3, // 0 for infinite, -1 for blocked

    @ColumnInfo(name = "snooze_count")
    val snoozeCount: Int = 0,

    @ColumnInfo(name = "smart_wake_enabled")
    val smartWakeEnabled: Boolean = false,

    @ColumnInfo(name = "smart_wake_window")
    val smartWakeWindowMinutes: Int = 20,

    @ColumnInfo(name = "has_habit_checklist")
    val hasHabitChecklist: Boolean = false,

    @ColumnInfo(name = "profile_id")
    val profileId: Int? = null,

    @ColumnInfo(name = "accountability_contact")
    val accountabilityContact: String? = null,

    @ColumnInfo(name = "accountability_delay")
    val accountabilityDelayMinutes: Int = 10,

    @ColumnInfo(name = "pre_alarm_enabled")
    val preAlarmEnabled: Boolean = false,

    @ColumnInfo(name = "pre_alarm_minutes")
    val preAlarmMinutes: Int = 5,

    @ColumnInfo(name = "adaptive_difficulty")
    val adaptiveDifficultyEnabled: Boolean = false,

    @ColumnInfo(name = "dismissal_history_count")
    val dismissalHistoryCount: Int = 0,

    @ColumnInfo(name = "last_modified")
    val lastModified: Long = System.currentTimeMillis()
) {
    fun isRepeating(): Boolean = repeatDays != 0

    /**
     * Calculates the next trigger time for this alarm in milliseconds.
     * Uses java.time APIs for DST-safe calculation.
     */
    fun getNextAlarmTime(): Long {
        val now = LocalDateTime.now()
        var scheduledTime = now.withHour(hour)
            .withMinute(minute)
            .withSecond(0)
            .withNano(0)

        // If the scheduled time is in the past, move to tomorrow
        if (scheduledTime.isBefore(now) || scheduledTime.isEqual(now)) {
            scheduledTime = scheduledTime.plusDays(1)
        }

        if (!isRepeating()) {
            return scheduledTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
        }

        // For repeating alarms, find the next active day
        for (i in 0 until 7) {
            // DayOfWeek is 1 (Monday) to 7 (Sunday)
            val dayOfWeek = scheduledTime.dayOfWeek.value
            val dayBit = 1 shl (dayOfWeek - 1)

            if ((repeatDays and dayBit) != 0) {
                return scheduledTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
            }
            scheduledTime = scheduledTime.plusDays(1)
        }

        // Fallback to the first occurrence if something goes wrong
        return scheduledTime.atZone(ZoneId.systemDefault()).toInstant().toEpochMilli()
    }
}
